hazelcast/connection.py

```diff
@@ -45,164 +112,404 @@ def add_listener(self, on_connection_opened=None, on_connection_closed=None):
-        uuid = self._client.cluster.uuid
-        owner_uuid = self._client.cluster.owner_uuid
-
-        request = client_authentication_codec.encode_request(
-            username=self._client.config.group_config.name,
-            password=self._client.config.group_config.password,
-            uuid=uuid,
-            owner_uuid=owner_uuid,
-            is_owner_connection=False,
-            client_type=CLIENT_TYPE,
-            serialization_version=SERIALIZATION_VERSION,
-            client_hazelcast_version=CLIENT_VERSION)
-
-        def callback(f):
-            parameters = client_authentication_codec.decode_response(f.result())
-            if parameters["status"] != 0:
-                raise AuthenticationError("Authentication failed.")
-            connection.endpoint = parameters["address"]
-            self.owner_uuid = parameters["owner_uuid"]
-            self.uuid = parameters["uuid"]
-            connection.server_version_str = parameters.get("server_hazelcast_version", "")
-            connection.server_version = calculate_version(connection.server_version_str)
+    def get_random_connection(self):
+        if self._smart_routing_enabled:
+            member = self._load_balancer.next()
+            if member:
+                connection = self.get_connection(member.uuid)
+                if connection:
+                    return connection
+
+        for connection in six.itervalues(self.active_connections):
             return connection
 
-        return self._client.invoker.invoke_on_connection(request, connection).continue_with(callback)
+        return None
 
-    def get_or_connect(self, address, authenticator=None):
-        """
-        Gets the existing connection for a given address. If it does not exist, the system will try to connect
-        asynchronously. In this case, it returns a Future. When the connection is established at some point in time, it
-        can be retrieved by using the get_connection(:class:`~hazelcast.core.Address`) or from Future.
+    def start(self):
+        if self.live:
+            return
 
-        :param address: (:class:`~hazelcast.core.Address`), the address to connect to.
-        :param authenticator: (Function), function to be used for authentication (optional).
-        :return: (:class:`~hazelcast.connection.Connection`), the existing connection or it returns a Future which includes asynchronously.
-        """
-        if address in self.connections:
-            return ImmediateFuture(self.connections[address])
+        self.live = True
+        self._load_balancer = self._client.load_balancer
+        self._heartbeat_manager.start()
+        self._connect_to_cluster()
+        if self._smart_routing_enabled:
+            self._start_connect_all_members_timer()
+
+    def shutdown(self):
+        if not self.live:
+            return
+
+        self.live = False
+        if self._connect_all_members_timer:
+            self._connect_all_members_timer.cancel()
+
```

Comment:
does `_connect_all_members_timer` wait for already running action to finish? 
If not we can end up with an open connection if I am not mistaken.

Reply:
We are canceling the timer. 
- If the timer is not started to execute yet, it won't be run.
- If the timer is executing right now, it will try to finish its execution(in the reactor thread, but we are not waiting for it right there). In the timer, before trying to `get_or_connect` to a member, we do have a `lifecycle.running` check. (which will be `False` after the shutdown call). If we can pass this check in the reactor thread, we can infact create a connection. We do have a cleanup code in the `reactor.shutdown` but it is cleaning up connection before executing the timers. (it may create connection after entering the loop in which we iterate over connections) Therefore, I moved waiting for timers to finish before cleaning up the connections. That should solve the problem